### ISSUE TITLE
UnicodeDecodeError if using AWS ElasticSearch cluster

### ISSUE THREAD

**Issue by StriveForBest** [wagtail/wagtail#3331]
### Issue Summary

Doesn't work with provided django settings. With alternative settings (from [Wagtail issue 2776](https://github.com/wagtail/wagtail/issues/2776#issuecomment-246412796), getting `UnicodeDecodeError` if any special character is present in `Page.title` (or any other indexable `Page` field I believe).

If using with AWS ElasticSearch service, first issue I came across is that default django settings from Wagtail documentation don't work if signing the request with `AWS4Auth` (recommended by AWS).

I found the solution that works in [Wagtail issue 2776](https://github.com/wagtail/wagtail/issues/2776#issuecomment-246412796). The very last comment by @justinoue.

But assuming I have a character like `u'C\xe9line'` (which is `u'Céline'`) as `Page.title`, `update_index` breaks with `UnicodeDecodeError`.

### Steps to Reproduce

1. Create `IAM` user, give it `AmazonESFullAccess - AWS Managed policy` permission.
2. Spin up AWS ElasticSearch instance with Access Policy allowing access to IAM user above.
3. Have any of the `Page` instances title set to `u'C\xc3\xa9line'`.
4. Use settings from Wagtail documentation:
```
from elasticsearch import RequestsHttpConnection
from requests_aws4auth import AWS4Auth

AWS_ELASTICSEARCH_ACCESS_KEY_ID = '<YOUR_ACCESS_ID>'
AWS_ELASTICSEARCH_SECRET_ACCESS_KEY = '<YOUR_SECRET_ACCESS_KEY>'

WAGTAILSEARCH_BACKENDS = {
     'default': {
         'BACKEND': 'wagtail.wagtailsearch.backends.elasticsearch2',
         'URLS': ['https://<AWS_ES_ENDPOINT>'],
         'INDEX': 'wagtail',
         'TIMEOUT': 5,
         'OPTIONS': {
             'connection_class': RequestsHttpConnection,
         },
         'INDEX_SETTINGS': {},
         'port': 443,
         'use_ssl': True,
         'verify_certs': True,
         'http_auth': AWS4Auth(AWS_ELASTICSEARCH_ACCESS_KEY_ID, AWS_ELASTICSEARCH_SECRET_ACCESS_KEY, '<AWS_ES_REGION>', 'es'),
     }
 }
```
5. ^^ This is going to break as it doesn't sign the requests and thinks you're trying to do this as anonymous user. The error looks like this:
```
elasticsearch.exceptions.AuthorizationException: TransportError(403, u'{"Message":"User: anonymous is not authorized to perform: es:ESHttpDelete on resource: am-dev"}')
```

6. Change the settings to match the example from [Wagtail issue 2776](https://github.com/wagtail/wagtail/issues/2776#issuecomment-246412796).
```
from elasticsearch import RequestsHttpConnection
from requests_aws4auth import AWS4Auth

AWS_ELASTICSEARCH_ACCESS_KEY_ID = '<YOUR_ACCESS_ID>'
AWS_ELASTICSEARCH_SECRET_ACCESS_KEY = '<YOUR_SECRET_ACCESS_KEY>'

WAGTAILSEARCH_BACKENDS = {
    'default': {
        'BACKEND': 'wagtail.wagtailsearch.backends.elasticsearch2',
        'INDEX': 'wagtail',
        'TIMEOUT': 5,
        'HOSTS': [{
            'host': '<AWS_ES_ENDPOINT>'',
            'port': 443,
            'use_ssl': True,
            'verify_certs': True,
            'http_auth': AWS4Auth(AWS_ELASTICSEARCH_ACCESS_KEY_ID, AWS_ELASTICSEARCH_SECRET_ACCESS_KEY, '<AWS_ES_REGION>', 'es'),
        }],
        'connection_class': RequestsHttpConnection
    }
}
```
7. Run `./manage.py update_index --verbosity=3`
8. You will get this error:
```
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 17345: ordinal not in range(128)
```

Full traceback is [here](https://gist.github.com/StriveForBest/1e5d926cf8f75b9e756ce200b777bbd8).

### Notes

1. `position 17345` where it breaks is the index of special character.
```
In [1]: message_body[17344:17351]
Out[1]: 'C\xc3\xa9line'
```
or specifically:
```
In [2]: message_body[17345]
Out[2]: '\xc3'
```
2. The settings in step 6 are not documented. Note how it uses `hosts` instead of `URLS` etc.
3. Interesting that the issue ultimately traces down to python's httplib.py:880 where it tries to do
```
msg += message_body
```
where `msg` is unicode but `message_body` is string containing our special character.

See [the gist](https://gist.github.com/StriveForBest/d2f410655b82aa2ebd03c839207a1edf) with variables `msg` and `message_body`. Note, some parameters from msg (`AWS_ES_ENDPOINT`, `SHA`, `AWS_ELASTICSEARCH_ACCESS_KEY_ID`, `SIGNATURE`) are hidden since they contain sensitive information. 

### Technical details

* Python version: `Python 2.7.13`
* Django version: `Django 1.9.6`
* Wagtail version: `Wagtail 1.8.1`

**Comment by StriveForBest** [comment]
@gasman just want to clarify on `some-day` milestone :)

Currently WagtailSearch is unusable with AWS ElasticSearch if any unicode characters are present. Just making sure it's not mistakenly tagged as non-important issue taken that AWS ES is one of the biggest services and is growing.

**Comment by gasman** [comment]
@StriveForBest The 'some-day' milestone simply means that within the core team there are currently no resources allocated towards fixing this - most likely because nobody has sufficient experience with AWS ElasticSearch - and so there's no ETA for the fix to be released. We're still interested in getting it fixed, but we're reliant on people like yourself in the wider Wagtail community to investigate and provide a fix...

**Comment by tomdyson** [comment]
p.s. @StriveForBest thank you for your exemplary issue report!

**Comment by StriveForBest** [comment]
@gasman / @tomdyson thanks guys, I understand the resourcing part of opensource work.

**Comment by gasman** [comment]
Hi @StriveForBest - according to https://github.com/wagtail/wagtail/issues/2776#issuecomment-324437853, it sounds like the correct fix is to put `'connection_class': RequestsHttpConnection` back inside the `OPTIONS` dict, so it becomes:

    WAGTAILSEARCH_BACKENDS = {
        'default': {
            'BACKEND': 'wagtail.wagtailsearch.backends.elasticsearch2',
            'INDEX': 'wagtail',
            'TIMEOUT': 5,
            'HOSTS': [{
                'host': '<AWS_ES_ENDPOINT>'',
                'port': 443,
                'use_ssl': True,
                'verify_certs': True,
                'http_auth': AWS4Auth(AWS_ELASTICSEARCH_ACCESS_KEY_ID, AWS_ELASTICSEARCH_SECRET_ACCESS_KEY, '<AWS_ES_REGION>', 'es'),
            }],
            'OPTIONS': {
                'connection_class': RequestsHttpConnection
            }
        }
    }

Can you confirm if this fixes the issue for you, please?

**Comment by gasman** [comment]
Actually, on closer inspection, I'm not sure if this will fix the issue - this is just the change required when upgrading to Wagtail 1.8 from an earlier version: http://docs.wagtail.io/en/v1.12/releases/1.8.html#extra-options-for-the-elasticsearch-constructor-should-be-now-defined-with-the-new-key-options-of-the-wagtailsearch-backends-setting

(However, if you were indeed using the above configuration against Wagtail 1.8.1, then that's something you should fix in any case!)

**Comment by dhl** [comment]
Hi @StriveForBest, have you tried using Elasticsearch 5? I'm currently using Wagtail 1.11.1 against an AWS Elasticsearch 5.3 cluster. Indexing works with accented characters.

Python version: `3.5.2`
Django version: `1.11.3`
Wagtail version: `1.11.1`
ElasticSearch version: `5.3`

**Comment by StriveForBest** [comment]
@dhl can you share your exact config? I'll try with my setup with Python 2 and both elasticsearch 2 and 5.

**Comment by dhl** [comment]
@StriveForBest my configuration is the same as what @gasman had posted above.

My guess is that when the fields we're grabbed for indexing, the encoding on the Python 2.7 byte strings weren't specified.

I wonder if changing the default encoding will work?

```python
import sys
reload(sys)
sys.setdefaultencoding('utf8')
```

This is obviously a nasty workaround, and it'd be better to fix the underlying problem, but maybe it'd help you run AWS Elasticsearch while the problem is being fixed.

**Comment by gasman** [comment]
Will close this issue now, as there haven't been any updates here for a while - things have moved on quite a bit since this was opened (the documentation fix in #3786, Elasticsearch 2 reaching end-of-life at the end of next month, and above all, the imminent dropping of Python 2 support), so it's unlikely that the setup described here will be applicable to current Elasticsearch installations. If there are any outstanding issues with AWS + Elasticsearch, they'll most likely be better served by opening a new issue.

**Comment by StriveForBest** [comment]
@gasman, makes sense. I will try a similar setup with Elasticsearch 5 and latest Wagtail soon and will open another issue if necessary. I think it's important to maintain Python 2 support.
